cpp/src/groupby/sort/sort_helper.cu

```diff
@@ -0,0 +1,312 @@
+    cudf::experimental::detail::scatter(
+      table_view({group_labels_view}), scatter_map, 
+      table_view({temp_labels->view()}),
+      false, rmm::mr::get_default_resource(), stream);
+
+  _unsorted_keys_labels = std::move(t_unsorted_keys_labels->release()[0]);
+
+  return _unsorted_keys_labels->view();
+}
+
+column_view helper::keys_bitmask_column(cudaStream_t stream) {
+  if (_keys_bitmask_column)
+    return _keys_bitmask_column->view();
+
+  auto row_bitmask = bitmask_and(_keys, rmm::mr::get_default_resource(), stream);
+
+  _keys_bitmask_column = make_numeric_column(
+    data_type(type_id::INT8), _keys.num_rows(), std::move(row_bitmask),
+    cudf::UNKNOWN_NULL_COUNT, stream);
+
+  using T = id_to_type<type_id::INT8>;
+  thrust::fill(rmm::exec_policy(stream)->on(stream),
+    _keys_bitmask_column->mutable_view().begin<T>(),
+    _keys_bitmask_column->mutable_view().end<T>(),
+    0);
+
+  return _keys_bitmask_column->view();
+}
+
+helper::column_ptr helper::sorted_values(column_view const& values, 
+  rmm::mr::device_memory_resource* mr,
+  cudaStream_t stream)
+{
+  column_ptr values_sort_order = cudf::experimental::detail::sorted_order(
+    table_view({unsorted_keys_labels(), values}), {},
+    std::vector<null_order>(2, null_order::AFTER), mr, stream);
+
+  // Zero-copy slice this sort order so that its new size is num_keys()
+  column_view gather_map = cudf::experimental::detail::slice(
+    values_sort_order->view(), 0, num_keys(stream) );
+
+  auto sorted_values_table = cudf::experimental::detail::gather(
+    table_view({values}), gather_map, false, false, false, mr, stream);
+
+  return std::move(sorted_values_table->release()[0]);
+}
+
+helper::column_ptr helper::grouped_values(column_view const& values, 
+  rmm::mr::device_memory_resource* mr,
+  cudaStream_t stream)
+{
+  auto gather_map = cudf::experimental::detail::slice(
+    key_sort_order(), 0, num_keys(stream));
+
+  auto grouped_values_table = cudf::experimental::detail::gather(
+    table_view({values}), gather_map, false, false, false, mr, stream);
+
+  return std::move(grouped_values_table->release()[0]);
+}
+
+std::unique_ptr<table> helper::unique_keys(
+  rmm::mr::device_memory_resource* mr,
+  cudaStream_t stream)
+{
+  auto idx_data = key_sort_order().data<size_type>();
+  auto transformed_group_ids = index_vector(num_groups());
+
+  thrust::transform(rmm::exec_policy(stream)->on(stream),
+                    group_offsets().begin(), group_offsets().end(),
+                    transformed_group_ids.begin(),
+    [idx_data] __device__ (size_type i) { return idx_data[i]; } );
+
+  auto gather_map = cudf::column_view(data_type(type_to_id<size_type>()),
+    num_groups(), transformed_group_ids.data().get());
+
+  // TODO (dm): replace this with iterator based gather when it's made so that 
+  //            gather_map is not required to be generated
+  return cudf::experimental::detail::gather(_keys, gather_map,
```

Comment:
Can't `cudf::unique` be called instead of a `gather`?

Reply:
Do you mean `drop_duplicates`. It uses the row comparator and I did some benchmarks which reveal a high cost to operations that use `row_equality_comparator`. Since offsets are already generated, it speeds up the operation.

Reply:
And yes, I do plan to look into the high cost of the comparators after this PR.

Reply:
Ah yeah, sure, that makes sense.